core/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/PutStoredScriptRequest.java

```diff
@@ -92,8 +97,27 @@ public BytesReference script() {
         return script;
     }
 
+    public XContentType xContentType() {
+        return xContentType;
+    }
+
+    /**
+     * Set the script source using bytes.
+     * @deprecated this method is deprecated as it relies on content type detection. Use {@link #script(BytesReference, XContentType)}
+     */
+    @Deprecated
     public PutStoredScriptRequest script(BytesReference source) {
         this.script = source;
+        this.xContentType = Objects.requireNonNull(XContentFactory.xContentType(source));
+        return this;
+    }
+
+    /**
+     * Set the script source and the content type of the bytes.
+     */
+    public PutStoredScriptRequest script(BytesReference source, XContentType xContentType) {
```

Comment:
I think I'll leave this one as is so we do not do a double parse of the script or move the ScriptMetaData logic into the request class (see the different ways stored scripts can be defined comment in the parseStoredScript method).